src/kbmod/search/raw_image_eigen.h

```diff
@@ -0,0 +1,250 @@
+
+      unsigned dx = right_x - left_x;
+      unsigned dy = bot_y - top_y;
+      return {left_x, dx, top_y, dy};
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const Index& rc);
+  };
+
+  std::ostream& operator<<(std::ostream& os, const Index& rc){
+    os << "x: " << rc.i << " y: " << rc.j;
+    return os;
+  }
+
+
+
+  struct Point{
+    float x;
+    float y;
+
+    Point(float xd, float yd)
+      : x(xd), y(yd) {}
+
+    Index to_index(){
+      return Index(x, y);
+    }
+
+    std::array<Point, 4> nearest_pixel_coords()  {
+      return {
+        Point(floor(x-0.5f)+0.5f, floor(y+0.5f)+0.5f),
+        Point(floor(x+0.5f)+0.5f, floor(y+0.5f)+0.5f),
+        Point(floor(x-0.5f)+0.5f, floor(y-0.5f)+0.5f),
+        Point(floor(x+0.5f)+0.5f, floor(y-0.5f)+0.5f)
+      };
+    }
+
+    std::array<Index, 4> nearest_pixel_idxs(){
+      return {
+        Index(x-0.5f, y+0.5f),
+        Index(x+0.5f, y+0.5f),
+        Index(x-0.5f, y-0.5f),
+        Index(x+0.5f, y-0.5f)
+      };
+    };
+
+    friend std::ostream& operator<<(std::ostream& os, const Point& rc);
+  };
+
+  std::ostream& operator<<(std::ostream& os, const Point& rc){
+    os << "x: " << rc.x << " y: " << rc.y;
+    return os;
+  }
+
+
+  // templates brother
+  using Image = Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>;
+  using ImageI = Eigen::Matrix<int, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>;
+  using ImageRef = Eigen::Ref<Image>;
+  using ImageIRef = Eigen::Ref<Image>;
+
+
+  class RawImageEigen {
+  public:
+    RawImageEigen();
+    explicit RawImageEigen(Image& img, double obs_time=-1.0d);
+    explicit RawImageEigen(unsigned w, unsigned h, float value=0.0f, double obs_time=-1.0d);
+
+    RawImageEigen(const RawImageEigen& old);  // Copy constructor
+    RawImageEigen(RawImageEigen&& source);    // Move constructor
+
+    RawImageEigen& operator=(const RawImageEigen& source);  // Copy assignment
+    RawImageEigen& operator=(RawImageEigen&& source);       // Move assignment
+
+    // Basic getter functions for image data.
+    unsigned get_width() const { return width; }
+    unsigned get_height() const { return height; }
+    unsigned get_npixels() const { return width * height; }
+
+    // Inline pixel functions.
+    float get_pixel(int x, int y) const {
+      return Index(x, y).is_within(width, height) ? image(x, y) : NO_DATA;
+    }
+
+    bool pixel_has_data(int x, int y) const {
+      return Index(x, y).is_within(width, height) ? image(x, y) != NO_DATA : false;
+    }
+
+    void set_pixel(int x, int y, float value) {
+      if (Index(x, y).is_within(width, height))
+        image(y, x) = value;
+    }
+
+    // Functions for locally storing the image time.
+    double get_obstime() const { return obstime; }
+    void set_obstime(double new_time) { obstime = new_time; }
+
+    // note, this produces a copy, because array is private
+    // it is not possible to return an usable reference to it.
```

Comment:
It's possible we could remove this function altogether and have the functions that use it access the pixels directly instead.

Reply:
Moved it, `Index.centered_block` still requires image size. Overall I think perhaps `Index` is a mistake and it should have been just a set of functions in the `indexing` namespace. For example `indexing::centered_block`, `indexing::to_indices(...., floor=true)` etc... Then we can make C++ inline by force if needed, correct?